src/github.com/getlantern/flashlight/ui/service.go

```diff
@@ -89,14 +114,29 @@ func read() {
 		}
 
 		// Delegating response to the service that registered with the given type.
-		if services[env.Type] != nil {
-			// Pass this message and continue reading another one.
-			go func() {
-				services[env.Type].In <- b
-			}()
-		} else {
-			log.Errorf("Message type %s belongs to an unkown service.", env.Type)
+		if services[envType.Type] == nil {
+			log.Errorf("Message type %v belongs to an unkown service.", envType.Type)
+			return
 		}
 
+		env := &envelope{}
+		err = json.Unmarshal(b, env)
+		if err != nil {
+			log.Errorf("Unable to unmarshal message of type %v: %v", envType.Type, err)
+			return
+		}
+		// Pass this message and continue reading another one.
+		services[env.Type].in <- env.Message
```

Comment:
>  the backend should back pressure to the UI rather than trying to buffer until it exhausts its own resources. That way, the UI will slow down (or start having errors), but the backend will remain stable and able to serve other consumers (e.g. user agents that are using the proxy).

That's a very important point if one wants to look for stability on server side.